File: benchmarks/baselines/python/performance_adapter.py
#!/usr/bin/env python3
"""Persistent M8 performance adapter for the frozen Python V2 boundary."""

from __future__ import annotations

import argparse
import json
import sys
import time
from pathlib import Path
from typing import Any

from v2.fixture import run_case


def _emit(value: dict[str, Any]) -> None:
    sys.stdout.write(
        json.dumps(value, ensure_ascii=False, separators=(",", ":")) + "\n"
    )
    sys.stdout.flush()


def _load_cases(manifest_path: Path) -> list[object]:
    manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
    return [
        json.loads(
            (manifest_path.parents[2] / entry["path"]).read_text(encoding="utf-8")
        )
        for entry in manifest["fixtures"]
    ]


def _run(cases: list[object]) -> list[dict[str, Any]]:
    return [run_case(case) for case in cases]


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--manifest", type=Path, required=True)
    args = parser.parse_args()
    cases = _load_cases(args.manifest.resolve())
    _emit({"type": "ready", "case_count": len(cases)})

    for line in sys.stdin:
        command = json.loads(line)
        kind = command["command"]
        if kind == "verify":
            _emit({"type": "verified", "results": _run(cases)})
        elif kind == "warmup":
            iterations = int(command["iterations"])
            checksum = 0
            for _ in range(iterations):
                checksum ^= sum(len(result["case_id"]) for result in _run(cases))
            _emit(
                {
                    "type": "warmed",
                    "iterations": iterations,
                    "request_count": iterations * len(cases),
                    "checksum": checksum,
                }
            )
        elif kind == "measure":
            duration_ns = int(command["duration_ns"])
            sample_stride = int(command["sample_stride"])
            started = time.perf_counter_ns()
            samples: list[int] = []
            request_count = 0
            checksum = 0
            while request_count == 0 or time.perf_counter_ns() - started < duration_ns:
                for case in cases:
                    before = time.perf_counter_ns()
                    result = run_case(case)
                    elapsed = time.perf_counter_ns() - before
                    if request_count % sample_stride == 0:
                        samples.append(elapsed)
                    request_count += 1
                    checksum ^= len(result["case_id"])
            elapsed_ns = time.perf_counter_ns() - started
            _emit(
                {
                    "type": "measured",
                    "clock": "time.perf_counter_ns",
                    "elapsed_ns": elapsed_ns,
                    "request_count": request_count,
                    "sample_stride": sample_stride,
                    "samples_ns": samples,
                    "checksum": checksum,
                }
            )
        elif kind == "shutdown":
            _emit({"type": "stopped"})
            return 0
        else:
            raise ValueError(f"unsupported command {kind!r}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: benchmarks/baselines/rust/v2/examples/performance_adapter.rs
//! Persistent M8 performance adapter for the frozen Rust V2 fixture boundary.

use std::{
    fs,
    io::{self, BufRead, Write},
    path::{Path, PathBuf},
    time::{Duration, Instant},
};

use ail_job_service_v2::codec::{run_case_value, SingleCaseResult};
use serde::Deserialize;
use serde_json::{json, Value};

#[derive(Deserialize)]
struct Manifest {
    fixtures: Vec<ManifestEntry>,
}

#[derive(Deserialize)]
struct ManifestEntry {
    path: String,
}

#[derive(Deserialize)]
struct Command {
    #[serde(rename = "command")]
    kind: String,
    iterations: Option<usize>,
    duration_ns: Option<u64>,
    sample_stride: Option<usize>,
}

fn emit(value: &Value) -> Result<(), Box<dyn std::error::Error>> {
    let stdout = io::stdout();
    let mut output = stdout.lock();
    serde_json::to_writer(&mut output, value)?;
    writeln!(output)?;
    output.flush()?;
    Ok(())
}

fn load_cases(path: &Path) -> Result<Vec<Value>, Box<dyn std::error::Error>> {
    let manifest: Manifest = serde_json::from_slice(&fs::read(path)?)?;
    let root = path
        .parent()
        .and_then(Path::parent)
        .and_then(Path::parent)
        .ok_or("manifest does not have a repository root")?;
    manifest
        .fixtures
        .iter()
        .map(|entry| {
            let content = fs::read(root.join(&entry.path))?;
            Ok(serde_json::from_slice(&content)?)
        })
        .collect()
}

fn run(cases: &[Value]) -> Result<Vec<SingleCaseResult>, Box<dyn std::error::Error>> {
    cases
        .iter()
        .cloned()
        .map(|value| Ok(run_case_value(value)?))
        .collect()
}

fn argument() -> Result<PathBuf, Box<dyn std::error::Error>> {
    let mut arguments = std::env::args().skip(1);
    while let Some(value) = arguments.next() {
        if value == "--manifest" {
            return arguments
                .next()
                .map(PathBuf::from)
                .ok_or_else(|| "missing --manifest value".into());
        }
    }
    Err("missing --manifest".into())
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let cases = load_cases(&argument()?)?;
    emit(&json!({"type": "ready", "case_count": cases.len()}))?;
    for line in io::stdin().lock().lines() {
        let command: Command = serde_json::from_str(&line?)?;
        match command.kind.as_str() {
            "verify" => {
                emit(&json!({"type": "verified", "results": run(&cases)?}))?;
            }
            "warmup" => {
                let iterations = command.iterations.ok_or("missing iterations")?;
                let mut checksum = 0;
                for _ in 0..iterations {
                    for result in run(&cases)? {
                        checksum ^= result.case_id.len();
                    }
                }
                emit(&json!({
                    "type": "warmed",
                    "iterations": iterations,
                    "request_count": iterations * cases.len(),
                    "checksum": checksum,
                }))?;
            }
            "measure" => {
                let duration =
                    Duration::from_nanos(command.duration_ns.ok_or("missing duration_ns")?);
                let stride = command.sample_stride.ok_or("missing sample_stride")?;
                let started = Instant::now();
                let mut samples = Vec::new();
                let mut request_count = 0;
                let mut checksum = 0;
                while request_count == 0 || started.elapsed() < duration {
                    for value in &cases {
                        let before = Instant::now();
                        let result = run_case_value(value.clone())?;
                        let elapsed = u64::try_from(before.elapsed().as_nanos())?;
                        if request_count % stride == 0 {
                            samples.push(elapsed);
                        }
                        request_count += 1;
                        checksum ^= result.case_id.len();
                    }
                }
                emit(&json!({
                    "type": "measured",
                    "clock": "std::time::Instant",
                    "elapsed_ns": u64::try_from(started.elapsed().as_nanos())?,
                    "request_count": request_count,
                    "sample_stride": stride,
                    "samples_ns": samples,
                    "checksum": checksum,
                }))?;
            }
            "shutdown" => {
                emit(&json!({"type": "stopped"}))?;
                return Ok(());
            }
            other => return Err(format!("unsupported command {other:?}").into()),
        }
    }
    Ok(())
}
